server/chat-plugins/rock-paper-scissors.ts

```diff
@@ -0,0 +1,406 @@
+		for (const id in this.playerTable) {
+			buf += Utils.html`<tr style="background-color: #6688AA"><td>${this.playerTable[id].name}</td>`;
+			buf += Utils.html`<td style="text-align: center">${this.points[id] || 0}</td></tr>`;
+		}
+		buf += `</table></div>`;
+		if (this.wins.length) {
+			buf += `<br />`;
+			for (const [i, entry] of this.wins.entries()) {
+				if (this.wins.length > 6) {
+					// we only wanna show the last 6 rounds
+					const diff = this.wins.length - 6;
+					if (i < diff) continue;
+				}
+				if (!entry) {
+					buf += `<div class="broadcast-red">Nobody won round ${i + 1}...</div>`;
+					continue;
+				}
+				const {name, choice} = entry;
+				buf += Utils.html`<div class="broadcast-green">${name} won round ${i + 1} with ${choice}!</div>`;
+			}
+		}
+		return buf;
+	}
+	sendScrollback() {
+		this.room.add(this.getScrollback()).update();
+	}
+	end() {
+		const [p1, p2] = Object.keys(this.playerTable).map(item => this.playerTable[item]);
+		this.addControls(`<h2>The game is over!</h2>`);
+		if (this.points[p1?.id] === this.points[p2?.id]) {
+			this.addField(Utils.html`Tie between ${p1} and ${p2}!`);
+			return this.startNextRound();
+		}
+		const winner = (this.points[p1.id] || 0) > (this.points[p2.id] || 0) ? p1 : p2;
+		const points = this.points[winner.id];
+		const message = Utils.html`<strong>${winner.name} won the game with ${Chat.count(points, 'points')}!</strong>`;
+		this.addField(message);
+		this.add(message);
+		this.sendFullLog();
+		this.destroy();
+	}
+	clearChoices() {
+		const [p1, p2] = this.players;
+		p1.currentChoice = "";
+		p2.currentChoice = "";
+	}
+	addControls(message: string) {
+		for (const id in this.playerTable) {
+			this.playerTable[id].sendRoom(`|controlshtml|<div class="pad">${message}</div>`);
+		}
+	}
+	getPlayers() {
+		return Object.keys(this.playerTable).map(item => this.playerTable[item]);
+	}
+	addField(message: string) {
+		return this.room.add(`${this.getScrollback()}<br /><div class="broadcast-green">${message}</div>`).update();
+	}
+	runMatch() {
+		const [p1, p2] = this.players;
+		const result = this.checkMatchup(p1.currentChoice, p2.currentChoice);
+		if (result === null) { // tie
+			this.add(`The players have tied! Nobody wins this round....`);
+			this.wins.push(null);
+		} else {
+			const winner = result === true ? p1 : p2;
+			this.add(Utils.html`${winner.name} wins the round! They gain a point.`);
+			this.addPoint(winner);
+			this.wins.push({
+				name: winner.name,
+				choice: winner.currentChoice,
+			});
+		}
+		if (this.currentRound === MAX_ROUNDS) {
+			this.add(`The game has hit the max number of rounds, and so will be ending.`);
+			return this.end();
+		}
+		this.clearChoices();
+		this.sendScrollback();
+		return this.startNextRound();
+	}
+	addPoint(player: RPSPlayer) {
+		const id = player.user.id;
+		if (!this.points[id]) this.points[id] = 0;
+		this.points[id]++;
+		return this.points[id];
```

Comment:
It looks like `points` should be a property of a player.

Reply:
This is the only place `player.user` is used, so that change would remove this problem.